File: lib/place_array.py
import numpy as np
from icecream import ic

"""
" This function computes the array locations according to the number of arrays
"""


def place_array(array, room):

    # Center coordinates of each array
    thetas = np.arange(0, 2 * np.pi, 2 * np.pi / array['N'])
    rho = array['R']
    array['centers'] = np.tile(room['dim'] / 2, array['N']).reshape((array['N'], 3))
    array['centers'][:, 0] = array['centers'][:, 0] + np.cos(thetas) * rho
    array['centers'][:, 1] = array['centers'][:, 1] + np.sin(thetas) * rho

    # Microphones' displacements within each array
    thetas = np.arange(0, 2 * np.pi, 2 * np.pi / array['micN'])
    rho = array['radius']
    array['positions'] = np.zeros((array['N'], array['micN'], 3))  # position of each mic in array
    for aa in range(array['N']):
        array['positions'][aa, :, :] = np.tile(array['centers'][aa], array['micN']).reshape((array['micN'], 3))
        array['positions'][aa, :, 0] = array['positions'][aa, :, 0] + np.cos(thetas) * rho
        array['positions'][aa, :, 1] = array['positions'][aa, :, 1] + np.sin(thetas) * rho

    return array

File: parametric_vm/lib/place_array.m
function array = place_array(array, room)
    %%PLACEARRAY
    % This function computes the array locations according to the number of arrays

    array.center = cell(array.N, 1);
    array.position = cell(array.N, 1);

    array_R = 1.7; % m
    [centerCoordinates(:,1), centerCoordinates(:,2)] = pol2cart(0:2*pi/array.N:2*pi*(1-1/array.N), array_R);
    centerCoordinates = centerCoordinates + room.dim(1:2)/2';

    [micCoordinate(:,1),micCoordinate(:,2)] = pol2cart(0:2*pi/array.micN:2*pi*(1-1/array.micN), array.radius);

    % Place the microphones in the environment.
    for aa = 1:array.N
        array.center{aa} = [centerCoordinates(aa,:),(room.z/2)];
        array.position{aa} = [micCoordinate,zeros(array.micN,1)]+ repmat(array.center{aa}, array.micN, 1);
    end

end
